core/src/main/scala/org/ensime/core/AnalyzerManager.scala

```diff
@@ -0,0 +1,362 @@
+    case Loaded => Loaded
+    case _ => NotLoaded
+  }
+}
+final case object Loaded extends FileStatus {
+  def append(status: FileStatus) = status match {
+    case Removed => Removed
+    case Unloaded => Unloaded
+    case _ => Loaded
+  }
+}
+final case object Removed extends FileStatus {
+  def append(status: FileStatus) = status match {
+    case Loaded => Loaded
+    case _ => Removed
+  }
+}
+final case object Unloaded extends FileStatus { // different from Removed
+  def append(status: FileStatus) = status match {
+    case Loaded => Loaded
+    case _ => Unloaded
+  }
+}
+
+final case class LoadedFilesData(statusOfFile: Map[SourceFileInfo, FileStatus], default: FileStatus)
+
+class AnalyzerManager(
+    broadcaster: ActorRef,
+    analyzerCreator: EnsimeProject => Props,
+    implicit val config: EnsimeConfig
+) extends Actor with ActorLogging with ModuleFinder with Stash {
+
+  // maps the active modules to their analyzers
+  private var analyzers: Map[EnsimeProject, ActorRef] = Map.empty
+  private val historyOfModule: mutable.Map[EnsimeProject, LoadedFilesData] = mutable.Map.empty
+  private val projectWideAnalyzer: Props = analyzerCreator(
+    EnsimeProject(
+      EnsimeProjectId("wholeProject", "compile"),
+      Seq(),
+      config.projects.flatMap(_.sources).toSet,
+      config.projects.flatMap(_.targets).toSet,
+      config.projects.flatMap(_.scalacOptions).distinct,
+      config.projects.flatMap(_.javacOptions).distinct,
+      config.projects.flatMap(_.libraryJars).toSet,
+      config.projects.flatMap(_.librarySources).toSet,
+      config.projects.flatMap(_.libraryDocs).toSet
+    )
+  )
+
+  override def preStart(): Unit = {
+    // initialize loadedFiles
+    config.projects foreach (p => historyOfModule += (p -> LoadedFilesData(Map.empty, NotLoaded)))
+    // legacy clients expect to see FullTypeCheckComlpeteEvent and AnalyzerReadyEvent on connection
+    broadcaster ! Broadcaster.Persist(AnalyzerReadyEvent)
+    broadcaster ! Broadcaster.Persist(FullTypeCheckCompleteEvent)
+  }
+
+  override def receive: Receive = ready
+
+  private def ready: Receive = withLabel("ready") {
+    case ReloadExistingFilesEvent =>
+      if (analyzers.isEmpty)
+        broadcaster ! AnalyzerReadyEvent
+      else
+        for {
+          (module, analyzer) <- analyzers
+        } analyzer forward ReloadExistingFilesEvent
+    case req: RpcAnalyserRequest =>
+      forwardToAnalyzer(req)
+  }
+
+  // helper method
+  private def getAnalyzerFor(module: EnsimeProject): ActorRef =
+    analyzers.get(module) match {
+      case Some(analyzer) => analyzer
+      case None =>
+        // spawn a new analyzer and attach it to module
+        val newAnalyzer = context.actorOf(analyzerCreator(module))
+        analyzers = analyzers + (module -> newAnalyzer)
+        newAnalyzer
+
+    }
+
+  def toFile(f: SourceFileInfo) = f.file match {
+    case RawFile(path) => path.toFile
+    case ArchiveFile(path, _) => path.toFile
+  }
```

Comment:
I though of adding `path` to `trait EnsimeFile` but I am not sure how to implement that for `ArchiveFile` though I think we'll have to use `fullPath`